### PR TITLE
realtek: clock support for RTL930X (v4)

### PR DESCRIPTION
**Pull request by plappermaul**
a clock driver to rule them all ...

### CODE REVIEW

File: target/linux/realtek/files-5.10/drivers/clk/realtek/Kconfig [hunk 1]
```diff
@@ -14,6 +14,7 @@ config COMMON_CLK_RTL83XX
 	  This driver adds support for the Realtek RTL83xx series basic clocks.
 	  This includes chips in the RTL838x series, such as RTL8380, RTL8381,
 	  RTL832, as well as chips from the RTL839x series, such as RTL8390,
-	  RT8391, RTL8392, RTL8393 and RTL8396.
+	  RT8391, RTL8392, RTL8393 and RTL8396 and chips of the RTL839X series
```
Comment (yarrick) [inline target/linux/realtek/files-5.10/drivers/clk/realtek/Kconfig:17]: chips of the RTL930X series?

Comment (plappermaul) [reply target/linux/realtek/files-5.10/drivers/clk/realtek/Kconfig]: fixed

**Comment by plappermaul** [comment]
Attached v2 of the PR with following changes

- unlocking of RTL838X registers
- Moved documentation to a new yaml file

**Comment by lucize** [comment]
boot tested with RTL9302B, not tested extensively

**Review by svanheule — CHANGES_REQUESTED** [review]
The commit "realtek: clock driver enhancement for RTL930X" does so many things at once, it's really not reviewable, nor are the changes useful to bisect later. Please split this into smaller patches that fix or add one issue.

Commit "realtek: switch RTL930X DT to new clock driver" then introduces new DT properties, which haven't been announced in any commit message before. I assume they are hidden in between the changes in the first patch.

Commit "realtek: fixes in RTL83XX DT" should come before the new features. The RTL83xx fixes should probably even be a separate PR, since the new features (RTL930x support) are less urgent IMHO.

Commit "realtek: DT documentation for clock driver" should remove the comment from the source file, and translate that into a DT binding in one commit. Now you remove the documentation in the first commit, but it's only added back at the very end of the series.

**Comment by plappermaul** [comment]
@svanheule Dont want to complain but dissecting and recompiling each and every patch for the updated PR with this slow build system just relocates the work.

**Comment by plappermaul** [comment]
Any complains left after the split?

**Comment by svanheule** [comment]
Sorry, haven't had the time to look at this properly again last week. I'll try to provide some comments this week.

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl839x-sram.S [hunk 2]
```diff
@@ -16,16 +16,17 @@
...
 .set	noreorder
 
 .globl	rtcl_839x_dram_start
 rtcl_839x_dram_start:
 
 /*
- * Functions start here and should avoid access to normal memory. REMARK! Do not forget about
- * stack pointer and dirty caches that might interfere.
+ * Functions start here and should avoid access to normal memory. REMARK! Do not
+ * forget about stack pointer and dirty caches that might interfere.
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl839x-sram.S:29]: Spurious change, not really relevant for this patch.

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c [hunk 3]
```diff
@@ -600,8 +592,12 @@ static struct clk_hw *rtcl_get_clkhw(struct of_phandle_args *clkspec, void *prv)
 
 static int rtcl_ccu_register_clocks(void)
 {
+	struct clk *clk;
 	int clk_idx, ret;
 
+	clk = clk_register_fixed_rate(NULL, "xtal_clk", NULL, 0, RTCL_XTAL_RATE);
+	clk_register_clkdev(clk, "xtal_clk", NULL);
+
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:678]: Aside from a simpler devicetree definition, what's the benefit of making the driver provide the fixed rate clock input?

In case of a dual-SoC system and synchronous ethernet support (which these chips do appear to support), it might be possible that the secondary SoC is not driven by an external crystal, but could be driven by some external clock generator that needs to be set up first. Having that in the devicetree would provide a proper clock hierarchy. Whereas hard-coding the clock input frequency here omits that explicit relation between the two.

Unless I am misunderstanding the point of this patch, I don't think omitting the 25MHz clock from the board definition (i.e. DTS) is the "right" thing to do. The clock/crystal is an external component, so it should be defined as such IMHO.

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c [hunk 4]
```diff
@@ -264,13 +259,13 @@ struct rtcl_round_set {
...
+		RTCL_ROUND_SET(300000000,	625000000,	25000000),
+		RTCL_ROUND_SET(200000000,	375000000,	25000000),
+		RTCL_ROUND_SET(100000000,	200000000,	25000000)
 	}, {
-		RTCL_ROUND_SET(400000000, 850000000, 25000000),
-		RTCL_ROUND_SET(100000000, 400000000, 25000000),
-		RTCL_ROUND_SET(50000000, 200000000, 50000000)
+		RTCL_ROUND_SET(400000000,	850000000,	25000000),
+		RTCL_ROUND_SET(100000000,	400000000,	25000000),
+		RTCL_ROUND_SET(50000000,	200000000,	50000000)
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:277]: These cosmetic changes aren't really related to making the 25MHz clock internal to the driver.

File: target/linux/realtek/files-5.10/Documentation/devicetree/bindings/clock/realtek,rtl8380-clock.yaml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+title: Realtek clock control unit Device Tree Bindings
+
+maintainers:
+  - Markus Stockhausen <markus.stockhausen@gmx.de>
+
+description: |
+  The driver provides basic clock support for the central core clock unit (CCU)
+  and its PLLs inside the RTL838X, RTL8389X and RTL930X SOC. All provided clocks
+  are derived from an oscillator that runs at a fixed speed of 25 MHz. Currently
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/Documentation/devicetree/bindings/clock/realtek,rtl8380-clock.yaml:15]: oscillator or external clock

File: target/linux/realtek/files-5.10/Documentation/devicetree/bindings/clock/realtek,rtl8380-clock.yaml [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  };
+ 
+  cpu_opp_table: opp-table-0 {
+      compatible = "operating-points-v2";
+      opp-shared;
+      opp00 {
+          opp-hz = /bits/ 64 <425000000>;
+      };
+      ...
+  }
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/Documentation/devicetree/bindings/clock/realtek,rtl8380-clock.yaml:50]: You should make this a second example:
- One simple example, with a minimal DT setup
- One extended example, with all possible options added.

That way, the DT tooling will be able to detect errors in your examples too.

Please also try to verify that your binding is valid: https://www.kernel.org/doc/html/latest/devicetree/bindings/writing-schema.html (I can't spot any obvious errors)

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c [hunk 7]
```diff
@@ -34,15 +34,13 @@
 #define RTCL_DDR2		2
 #define RTCL_DDR3		3
 
-#define REG_CTRL0		0
-#define REG_CTRL1		1
-#define REG_COUNT		2
+#define RTCL_REGCNT		2
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:38]: Can you put this in the other patch with prefix changes?

And REG_CTRL{0,1} weren't used then?

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c [hunk 8]
```diff
@@ -74,7 +72,7 @@ static const int rtcl_regs[RTCL_SOCCNT][REG_COUNT][CLK_COUNT] = {
 	}
 
 struct rtcl_reg_set {
-	unsigned int rate;
+	unsigned long rate;
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:92]: Why the change? Isn't (unsigned) int the same size as (unsigned) long anyway?

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c [hunk 9]
```diff
@@ -1,4 +1,4 @@
-// SPDX-License-Identifier: GPL-2.0-only
+/* SPDX-License-Identifier: GPL-2.0-only */
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:1]: SPDX identifiers with `//` are correct in a source file.
See: https://www.kernel.org/doc/html/v4.18/process/license-rules.html#license-identifier-syntax

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c [hunk 10]
```diff
@@ -238,9 +239,9 @@ struct rtcl_clk {
 };
 
 static const struct rtcl_clk_info rtcl_clk_info[CLK_COUNT] = {
-	RTCL_CLK_INFO(CLK_CPU, "cpu_clk", "xtal_clk", "CPU"),
-	RTCL_CLK_INFO(CLK_MEM, "mem_clk", "xtal_clk", "MEM"),
-	RTCL_CLK_INFO(CLK_LXB, "lxb_clk", "xtal_clk", "LXB")
+	RTCL_CLK_INFO(CLK_CPU, "cpu_clk", "xtal_clk", "xtal_clk", "CPU"),
+	RTCL_CLK_INFO(CLK_MEM, "mem_clk", "xtal_clk", "xtal_clk", "MEM"),
+	RTCL_CLK_INFO(CLK_LXB, "lxb_clk", "xtal_clk", "xtal_clk", "LXB")
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c]: Why have two identical names?

This patch doesn't seem to line up with the commit message either... If a clock may not necessarily have a fixed clock as source, why not just use the framework by putting everything in the DTS?

File: target/linux/realtek/files-5.10/include/dt-bindings/clock/rtl83xx-clk.h [hunk 11]
```diff
@@ -7,9 +7,10 @@
...
 
-#define CLK_CPU		0
-#define CLK_MEM		1
-#define CLK_LXB		2
-#define CLK_COUNT	3
+#define CLK_SW		0
+#define CLK_CPU		1
+#define CLK_MEM		2
+#define CLK_LXB		3
+#define CLK_COUNT	4
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/include/dt-bindings/clock/rtl83xx-clk.h:14]: Why are you changing the values of these `#define`s? Adding `#define CLK_SWITCH 4` would be a smaller change, and the values are arbitrary anyway.

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h [hunk 12]
```diff
@@ -14,18 +14,29 @@
 #define RTL838X_PLL_GLB_CTRL			(0x0fc0)
 #define RTL838X_PLL_CPU_CTRL0			(0x0fc4)
 #define RTL838X_PLL_CPU_CTRL1			(0x0fc8)
+#define RTL838X_PLL_CPU_MISC_CTRL		(0x0fcc)
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h:17]: I don't see the `*_MISC_CTRL` registers being used here. I suppose it doesn't hurt to have the defined, but they are sort of sneaked in with another change.

File: target/linux/realtek/dts-5.10/rtl838x.dtsi [hunk 13]
```diff
@@ -62,17 +62,9 @@
...
-		compatible = "fixed-clock";
-		#clock-cells = <0>;
-		clock-frequency = <25000000>;
-	};
-
 	ccu: clock-controller {
 		compatible = "realtek,rtl8380-clock";
 		#clock-cells = <1>;
-		clocks = <&osc>;
-		clock-names = "ref_clk";
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl838x.dtsi:75]: I don't think being able to omit this, is worth the extra boilerplate in the clock driver.

File: target/linux/realtek/dts-5.10/rtl838x.dtsi [hunk 14]
```diff
@@ -262,10 +254,16 @@
...
 		#address-cells = <1>;
 		#size-cells = <1>;
-		ranges = <0 0x9f000000 0x10000>;
+		ranges = <0 0x9f000000 0x20000>;
+
+		sram_scratch: scratch-sram@0 {
+			compatible = "realtek,scratch";
+			reg = <0x0 0x20>;
+			pool;
+		};
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl838x.dtsi:266]: Any reason this has/needs a specific compatible?

File: target/linux/realtek/dts-5.10/rtl838x.dtsi [hunk 15]
```diff
@@ -262,10 +254,16 @@
 
 	sram0: sram@9f000000 {
 		compatible = "mmio-sram";
-		reg = <0x9f000000 0x10000>;
+		reg = <0x9f000000 0x20000>;
 		#address-cells = <1>;
 		#size-cells = <1>;
-		ranges = <0 0x9f000000 0x10000>;
+		ranges = <0 0x9f000000 0x20000>;
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl838x.dtsi:260]: Separate commit with Fixes: tag? Then that can be merged ahead of some of the bigger changes.

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c [hunk 16]
```diff
@@ -247,12 +314,16 @@ struct rtcl_clk {
...
+ */
 static const struct rtcl_clk_info rtcl_clk_info[CLK_COUNT] = {
-	RTCL_CLK_INFO(CLK_SW, "sw_clk", "xtal_clk", "xtal_clk", "SW"),
-	RTCL_CLK_INFO(CLK_CPU, "cpu_clk", "xtal_clk", "xtal_clk", "CPU"),
-	RTCL_CLK_INFO(CLK_MEM, "mem_clk", "xtal_clk", "xtal_clk", "MEM"),
-	RTCL_CLK_INFO(CLK_LXB, "lxb_clk", "xtal_clk", "xtal_clk", "LXB")
+	RTCL_CLK_INFO(CLK_SW, "sw_clk", "xtal_clk", "xtal_clk", "xtal_clk", "SW"),
+	RTCL_CLK_INFO(CLK_CPU, "cpu_clk", "xtal_clk", "xtal_clk", "xtal_clk", "CPU"),
+	RTCL_CLK_INFO(CLK_MEM, "mem_clk", "xtal_clk", "xtal_clk", "xtal_clk", "MEM"),
+	RTCL_CLK_INFO(CLK_LXB, "lxb_clk", "xtal_clk", "xtal_clk", "sw_clk", "LXB")
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:326]: Aha, this is why you have all those duplicated parent clocks. You should be able to encode this in the DTS.

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h [hunk 17]
```diff
@@ -79,12 +94,25 @@
...
+#define RTL930X_PLL_MEM_CTRL3_CMU_NCODE_IN(v)	(((v) >> 24) & 0xff)
+
 /*
- * Other stuff
+ * SRAM stuff
  */
 
 #define RTL_SRAM_MARKER				(0x5eaf00d5)
 #define RTL_SRAM_BASE_CACHED			(0x9f000000)
+#define RTL_SRAM_BASE_UNCACHED			(0xbf000000)
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h:118]: You've removed the `sync` ops earlier, shouldn't you use uncached memory there as well then?

**Review by svanheule — CHANGES_REQUESTED** [review]
Please avoid abbreviations like "reorg" in you commit titles/messages. When you're going over the commit message, also have a look at capitalising the sentences, and adding full stops at the end.

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c [hunk 18]
```diff
@@ -484,6 +481,26 @@ static const struct clk_ops rtcl_clk_ops = {
...
+ * Some RTL838X devices are shipped with register access locked. In this case
+ * we cannot read and/or write to LXB or SW PLL registers. As there is no real
+ * benefit of always unlocking/locking these registers just open up everything.
+ */
+	enable = RTL838X_INT_RW_CTRL_READ_EN | RTL838X_INT_RW_CTRL_WRITE_EN;
+	reg = ioread32((void *)RTL838X_INT_RW_CTRL);
+	if ((reg & enable) != enable) {
+		reg |= enable;
+		iowrite32(reg, (void *)RTL838X_INT_RW_CTRL);
+		pr_warn("%s: registers unlocked\n", __func__);
```
[the remark this answers is not in the document]
Comment (plappermaul) [reply target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:580]: Why should I warn about unlocked registers? That is the desired state and if set correctly by someone else (vendor) everything is fine. Opening up the registers should have a meaningful level.

File: target/linux/realtek/files-5.10/Documentation/devicetree/bindings/clock/realtek,rtl8380-clock.yaml (hunk 5 above)
Comment (plappermaul) [reply target/linux/realtek/files-5.10/Documentation/devicetree/bindings/clock/realtek,rtl8380-clock.yaml:15]: datasheet states: "25MHz Crystal Clock"

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c (hunk 8 above)
Comment (plappermaul) [reply target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:92]: wherever rate is used we take the long type. So be consistent here.

File: target/linux/realtek/dts-5.10/rtl838x.dtsi (hunk 13 above)
Comment (plappermaul) [reply target/linux/realtek/dts-5.10/rtl838x.dtsi:75]: This is only needed if oscillator stays in DT as fixed clock

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c (hunk 16 above)
Comment (plappermaul) [reply target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:326]: Dependency SW/LXB cannot be encoded in the DT as both are provided by the driver itself.

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h (hunk 17 above)
Comment (plappermaul) [reply target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h:118]: sync is a no longer needed leftover from c compiled sources that I copy/pasted to assembler.

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (svanheule) [reply target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:580]: That's exactly my point. "unlocked registers" is the desired state, so that's what you set. "locked registers" is just the default state, so not necessarily an indication that something is wrong, which would warrant a `pr_warn()`. I don't think you need to inform the user at all, just unlock the registers and be done with it.

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c (hunk 16 above)
Comment (svanheule) [reply target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:326]: I think you should be able to self-reference the clock node:
```
ccu: clock-controller {
	compatible = "realtek,rtl8380-clock";
	#clock-cells = <1>;
	clocks = <&osc>, <&ccu CLK_SW>;
	clock-names = "ref_clk", "sw_clk";
};

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h (hunk 17 above)
Comment (svanheule) [reply target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.h:118]: Why are they no longer needed though? I don't recall reading that in the commit message.

File: target/linux/realtek/dts-5.10/rtl838x.dtsi (hunk 14 above)
Comment (svanheule) [reply target/linux/realtek/dts-5.10/rtl838x.dtsi:266]: OK, this is something that would go into Documentation/devicetree/bindings/sram/sram.yaml then. Maybe add a label too.; something that indicates this is used by the clock driver to update the PLL registers: e.g.`label = "clock-config-update";` 

Can you use `protect-exec;` or is that not possible when also writing the blob from the driver?

File: target/linux/realtek/files-5.10/Documentation/devicetree/bindings/clock/realtek,rtl8380-clock.yaml (hunk 5 above)
Comment (svanheule) [reply target/linux/realtek/files-5.10/Documentation/devicetree/bindings/clock/realtek,rtl8380-clock.yaml:15]: image
"When using a crystal", so might also be an external clock driver.

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c (hunk 8 above)
Comment (svanheule) [reply target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:92]: OK, it appears you changing for consistency with the clock framework, which uses "unsigned long" for clock rates. Please try to put these "why"s in the commit message, so we don't have to spend time wondering why you're making changes like this.

**Comment by plappermaul** [comment]
Hi Sander,

maybe not the perfect place, but best I can see. While I appreciate your reviews and feedback I get the impression that developing new stuff for devices and pushing it to OpenWrt is not the favourite way to go. Why? Most of the feedback I got until now was of formal nature. This is not a problem as long as it does not get into a chair cycle discussion for that spare time is too valuable. Examples?

- A fixed frequency clock inside the driver is not favoured. Mainline kernel has >100 of these.
- In v1 a DT documentation is encouraged, v2 provides it (not much I can see elsewhere in OpenWrt) and still this is not good enough.
- changing variable types/names should be explained in commit messages
- a 7 line commit changing a timeout loop results in a discussion about capitalisation an where/how to reference commits
- The CAMEOTAG patch came close to a mathematical lecture

The first driver version went over 9 iterations and more than 200 comments until it got accepted. No problem for the initial round. But looking at the current feedback for 17 small patches for v2 we might top the numbers by far. Just for cleaning things up and adding support for a new device inside the driver?

To my personal pipeline (irq driver fix, 930x SMP activation, the new timer driver, 5.15 testing) I added the first stuff for implementing a hw accellerated AES/SHA1 crypto driver for the engine inside the 838x and 930x. With hours invested in understing asynchronous hash/cipher handling I'm thinking about the discussions and must-have-it-that-way-requests that will start afterwards.

To quote @robimarko from the old thread "the maintainers opinion pretty much equates to the project's opinion". I agree with that 100%. With that in mind the 5.15 RFC discussion "don't want to merge with L2 learning disable either, ... <he> wasn't able to figure out yet where things go wrong, but hopefully someone else can." feels strange. 

Writing this 40 days after Birger left seems hard but a good point for thinking how to go on from here. Development by proxy feels odd. And so it does whenever someone is complaining about old bad code and in the same sentence is setting highest expectations for new changes. I prefer reasonable patches that might not be perfect but keep it rolling.

**Comment by svanheule** [comment]
Hi Markus,

As much as I would like to do development myself, I currently do not have the time do so, since this is a spare time thing for me. That means "development by proxy" is what I am currently relegated to. I realize that my comment on the 5.15 issue can come across as "you guys do all the hard work, I'll just sit back and watch". That was not what I meant to say, so I must apologize. Hiroshi had these changes set aside for a while, and we would need to make the transition at some point. I figured a pull request may get it the extra sets of eyes it needed to pass the final hurdles.

Big changes like this patch set are hard to review, because they do so many things at once. I personally find it discouraging to have to pour through change dumps, and we *have* needed to hunt down bugs introduced by big, badly explained change sets before. So I really appreciate that you took the time and effort to split the changes into smaller patches. Now I think I have a better idea of the changes you want to make. Most changes are just fine, by the way, but I commented with most things that came to mind, both major and minor, so we don't end up at v10 before I start with the formalities.

Even if things aren't perfectly to your liking the first time around, I think it can sometimes be preferable to keep them that way if possible, to not distract from the actual changes you need to make to go forward. For example, the devicetree binding is a hard requirement upstream, but I am of the opinion that here, for OpenWrt, a comment in the source is sufficient. Yes, a binding will be required, but I am happy to keep things as they are here. My apologies for not making that sufficiently clear. If you want to include a proper binding with the other changes now, I won't object to that, but it also means I will take (and need) time to look at them and give feedback.

I am happy to see more people, like you, providing new code. As much as I can, I try to understand the code you are providing, and new features they expose. As you noticed, that means certain choices you make may pass the first time, because it's impossible for me to understand every little detail of a completely new driver, in a subsystem I've never used before (which means the majority of subsystems). When you make changes afterwards, I am left wondering why things need to change if I don't see the "why" of a change in the commit message.

What I *try* to do as a maintainer (but certainly do not always succeed at), is to keep things consistent among different developers. You have your preferences, and so does Olliver, or anybody else contributing to the project. In my opinion, consistency is key to also enable others to contribute back to the project. So yes, I fear that will inevitably include things that may come across as mere formalities and nit-picking. While I agree that this can be frustrating, once code is in OpenWrt (or the upstream kernel for example), it becomes part of the project. I'm afraid we can't just have the project be somebody's personal playground, just to give some code more exposure.

Please don't get discouraged by the process. I do appreciate your contributions to the project, and hope you decide to keep contributing, even if things are slow sometimes.

**Comment by plappermaul** [comment]
To get forward here I'll take your word for it "Most changes are just fine". So I have come to the following lonely descision.

- I changed the commit messages to better reflect their content.
- I did not (and will not) change the content. I do not like to rearrange 17 commits just to come to the same result.
- After your better understanding of the code, I ask you to have a quick look at the complete final package (at least C file).
- If something really should get rework (and is worth the effort), drop a note here or in the chat for me.
- I will fix it in a follow-up

This will keep maintainer & developer overhead small. So feel free to merge.

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c [hunk 19]
```diff
@@ -484,6 +481,26 @@ static const struct clk_ops rtcl_clk_ops = {
...
+
+	if (rtcl_ccu->soc != RTCL_SOC838X)
+		return;
+/*
+ * Some RTL838X devices are shipped with register access locked. In this case
+ * we cannot read and/or write to LXB or SW PLL registers. As there is no real
+ * benefit of always unlocking/locking these registers just open up everything.
+ */
+	enable = RTL838X_INT_RW_CTRL_READ_EN | RTL838X_INT_RW_CTRL_WRITE_EN;
+	reg = ioread32((void *)RTL838X_INT_RW_CTRL);
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:576]: This crashes my device on boot, since it's missing the switch core address space offset.

File: target/linux/realtek/dts-5.10/rtl838x.dtsi (hunk 14 above)
Comment (svanheule) [reply target/linux/realtek/dts-5.10/rtl838x.dtsi:266]: Unless I'm missing something, these new sram-scratch nodes also aren't used anywhere. Why are you adding them? If you need them for an upcoming feature, these should be included with those patches.

File: target/linux/realtek/dts-5.10/rtl930x.dtsi [hunk 20]
```diff
@@ -8,14 +10,57 @@
...
-		frequency = <800000000>;
 
 		cpu@0 {
 			compatible = "mips,mips34Kc";
 			reg = <0>;
+			clocks = <&ccu CLK_CPU>;
+			operating-points-v2 = <&cpu_opp_table>;
+		};
+
+		cpu@1 {
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl930x.dtsi:29]: Same comment as for the sram-scratch nodes. If this is still unused at this point, please leave it out of the change set. AFAICT, adding the second VPE is not required for switching over to the new clock driver; which this commit is supposedly doing.

File: target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c (hunk 19 above)
Comment (plappermaul) [reply target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c:576]: fixed, thanks for testing

File: target/linux/realtek/dts-5.10/rtl930x.dtsi (hunk 20 above)
Comment (plappermaul) [reply target/linux/realtek/dts-5.10/rtl930x.dtsi]: removed

File: target/linux/realtek/dts-5.10/rtl838x.dtsi (hunk 14 above)
Comment (plappermaul) [reply target/linux/realtek/dts-5.10/rtl838x.dtsi:266]: They are used in the 930x assembler code (SRAM_BASE_UNCACHED) to "reset" SRAM communication. References in other DTs have been removed.

**Comment by plappermaul** [comment]
updated as requested.

**Comment by plappermaul** [comment]
@svanheule maintainer ping after 11 days.

**Comment by plappermaul** [comment]
@svanheule Another 11 days over without a single bit of feedback. A simple "will not merge because of ..." will be enough.

**Comment by robimarko** [comment]
I am sure he is just really busy, OpenWrt rarely pays bills for devs

**Comment by plappermaul** [comment]
rebased

**Comment by plappermaul** [comment]
> I am sure he is just really busy, OpenWrt rarely pays bills for devs

Same for me.
